File: Python/isValid.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: isValid.py
@Function: 有效的括号
@Link: https://leetcode-cn.com/problems/valid-parentheses/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-07-25
"""


class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) % 2 == 1:
            return False

        pairs = {
            ")": "(",
            "]": "[",
            "}": "{",
        }
        stack = list()
        for ch in s:
            if ch in pairs:
                if not stack or stack[-1] != pairs[ch]:
                    return False
                stack.pop()
            else:
                stack.append(ch)

        return not stack


# --------------------------------------
if __name__ == "__main__":
    # s = "()[]{}"
    s = "([)]"

    solution = Solution()
    check_valid = solution.isValid(s)
    print(f"The solution of this problem is {check_valid}")

File: C++/isValid.cpp
/**
 * @File    : isValid.cpp
 * @Brief   : 有效的括号
 * @Link    : https://leetcode-cn.com/problems/valid-parentheses/
 * @Author  : Wei Li
 * @Date    : 2021-07-25
*/

#include <iostream>
#include <string>
#include <stack>

class Solution
{
public:
    bool isValid(std::string s)
    {
        std::stack<char> parsed;
        for (int i = 0; i < s.length(); ++i)
        {
            if (s[i] == '{' || s[i] == '[' || s[i] == '(')
            {
                parsed.push(s[i]);
            }
            else
            {
                if (parsed.empty())
                {
                    return false;
                }
                char c = parsed.top();
                if ((s[i] == '}' && c == '{') ||
                    (s[i] == ']' && c == '[') ||
                    (s[i] == ')' && c == '('))
                {
                    parsed.pop();
                }
                else
                {
                    return false;
                }
            }
        }
        return parsed.empty();
    }
};

// -----------------------------
int main(int argc, char **argv)
{
    // std::string str = "()";
    std::string str = "([)]";

    Solution solution;
    bool check_valid = solution.isValid(str);
    std::cout << "The solution of this problem is " << std::boolalpha << check_valid << std::endl;

    return 0;
}
